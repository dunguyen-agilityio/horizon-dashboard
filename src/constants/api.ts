export const API_ENTITY = {
  COMPLEX_REPORT: 'complex-report',
  CHECK_REPORT: 'check-report',
  FINANCIAL_REPORT: 'financial-report',
  USERS: 'users',
  SIGN_IN: 'auth/local',
  SIGN_UP: 'auth/local/register',
  FORGET_PASSWORD: 'auth/forgot-password',
  RESET_PASSWORD: 'auth/reset-password',
  CONTRIBUTOR: 'contributors',
};
