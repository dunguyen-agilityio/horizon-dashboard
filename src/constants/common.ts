export const QUERY_KEY = 'q';
export const STEP_KEY = 'step';
