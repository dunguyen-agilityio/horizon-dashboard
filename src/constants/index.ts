export * from './common';
export * from './images';
export * from './messages';
export * from './routes';
export * from './text';
export * from './endpoint';
export * from './session';
