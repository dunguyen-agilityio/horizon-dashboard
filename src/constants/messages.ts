export const TASK_MESSAGE = {
  MIN_LENGTH: 'Title is too short. Please enter at least 6 characters.',
};
