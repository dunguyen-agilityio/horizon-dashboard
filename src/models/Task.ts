import { TUser } from './User';

export const enum LABEL {
  PENDING = 'Pending',
  UPDATES = 'Updates',
  DONE = 'Done',
  ERRORS = 'Errors',
}

export const enum STATUS {
  BACKLOG = 'backlog',
  IN_PROGRESS = 'in-progress',
  DONE = 'done',
}

export class Task {
  id!: string;
  title!: string;
  description!: string;
  createdAt!: Date | string;
  updatedAt!: Date | string;
  dueDate?: string;
  startedDate?: string;
  status!: STATUS;
  labels!: LABEL[];
  assignees?: TUser[];
  cover?: string;

  constructor(task?: Task) {
    const {
      title = '',
      description = '',
      createdAt,
      updatedAt,
      dueDate,
      startedDate,
      assignees = [],
      labels = [],
    } = task || {};

    Object.assign(this, {
      ...task,
      title,
      description,
      assignees,
      labels: labels.length ? labels : [LABEL.PENDING],
      dueDate: dueDate ?? '',
      createdAt: createdAt ? new Date(createdAt) : undefined,
      updatedAt: updatedAt ? new Date(updatedAt) : undefined,
      startedDate: startedDate ?? '',
    });
  }
}
