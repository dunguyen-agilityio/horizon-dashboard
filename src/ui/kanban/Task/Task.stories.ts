import type { Meta, StoryObj } from '@storybook/react';

import { Task } from '.';

import { MOCK_TASKS } from '@/mocks/task';

const mockTask = MOCK_TASKS[0];
const mockCover = '/default-task-cover.webp';

const meta = {
  title: 'Components/Task',
  component: Task,
  tags: ['autodocs'],
} satisfies Meta<typeof Task>;

export default meta;
type Story = StoryObj<typeof meta>;

export const TaskMissCover: Story = {
  args: { ...mockTask, cover: undefined },
};

export const TaskWithCover: Story = {
  args: { ...mockTask, cover: mockCover },
};
