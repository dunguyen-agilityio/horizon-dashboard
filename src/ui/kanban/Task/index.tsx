// Components
import Image from 'next/image';
import Pen from '@/icons/Pen';
import Text from '@/components/Text';
import { Chip } from '@nextui-org/chip';
import Button from '@/components/Button';
import { Avatar, AvatarGroup } from '@nextui-org/avatar';

// Types
import { TEXT_SIZE, TEXT_VARIANT } from '@/types/text';

// Models
import { Task as TaskModel } from '@/models/Task';

// Utils
import { getColorByLabel } from '@/utils/task';

export const Task = ({
  title,
  description,
  labels,
  cover,
  assignees,
}: Pick<
  TaskModel,
  'title' | 'description' | 'labels' | 'cover' | 'assignees'
>) => (
  <div className="flex flex-col gap-[10px] pt-[18px] pl-[23px] pb-[23px] pr-[21px] rounded-[15px] bg-white dark:bg-indigo-light max-w-[466px]">
    <div className="flex justify-between">
      <Text size={TEXT_SIZE.lg} as="h3" className="max-w-[90%]">
        {title}
      </Text>
      <Button size="fit" aria-label="edit-button" className="bg-transparent">
        <Pen />
      </Button>
    </div>
    {cover && (
      <div className="relative min-h-[285px] min-w-[308px]">
        <Image src={cover} alt="cover" fill />
      </div>
    )}
    <Text as="p" variant={TEXT_VARIANT.SECONDARY}>
      {description}
    </Text>
    <div className="flex justify-between mt-5 items-center">
      <AvatarGroup isBordered max={3}>
        {assignees?.map(({ id, avatar, fullName }) => (
          <Avatar
            key={id}
            src={avatar}
            ImgComponent={Image}
            alt={fullName}
            imgProps={{ width: 20, height: 20 }}
            isBordered
          />
        ))}
      </AvatarGroup>
      <Chip
        color={getColorByLabel(labels[0])}
        className="text-white uppercase text-xs leading-[15px] rounded-[10px] py-[6.5px] px-[25px]"
      >
        {labels[0]}
      </Chip>
    </div>
  </div>
);
