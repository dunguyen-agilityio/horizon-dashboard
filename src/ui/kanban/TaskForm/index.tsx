'use client';

// Libs
import { useTransition } from 'react';
import { Controller, useForm } from 'react-hook-form';

// Components
import Button from '@/components/Button';
import { Input } from '@nextui-org/input';
import { Spinner } from '@nextui-org/spinner';

// Types
import { STATUS } from '@/models/Task';
import { type TaskFormData } from '@/types/task';

// Constants
import { MESSAGES } from '@/constants/messages';

interface TaskFormProps {
  status: STATUS;
  onClose: () => void;
  onSubmit: (data: TaskFormData) => Promise<void>;
}

const TaskForm = ({ onClose, onSubmit, status }: TaskFormProps) => {
  const [isPending, startTransition] = useTransition();

  const {
    control,
    handleSubmit,
    formState: { isDirty },
  } = useForm<TaskFormData>({
    mode: 'onBlur',
    defaultValues: { title: '', status },
  });

  const handleSubmitData = handleSubmit(async (data: TaskFormData) => {
    startTransition(async () => {
      await onSubmit(data);
    });
  });

  return (
    <div className="relative pt-[18px] pl-[23px] pb-[23px] pr-[21px] rounded-[15px] bg-white shadow-light-card dark:shadow-none dark:bg-indigo-light max-w-[466px]">
      <form
        onSubmit={handleSubmitData}
        className="flex flex-col gap-4"
        data-testid="add-new-task-form"
      >
        <Controller
          control={control}
          name="title"
          rules={{
            minLength: {
              message: MESSAGES.TASK.MIN_LENGTH,
              value: 6,
            },
          }}
          render={({ field, fieldState: { error, invalid } }) => (
            <Input
              {...field}
              label="Title"
              labelPlacement="outside"
              placeholder="Enter title"
              isInvalid={invalid}
              errorMessage={error?.message}
              size="md"
              data-testid="title-field"
            />
          )}
        />
        <div className="flex justify-between">
          <Button
            variant="faded"
            type="submit"
            isDisabled={!isDirty}
            className="bg-transparent"
          >
            {isPending ? <Spinner size="sm" color="success" /> : 'Create task'}
          </Button>
          <Button variant="faded" onClick={onClose} className="bg-transparent">
            Cancel
          </Button>
        </div>
      </form>
      {isPending && (
        <div className="absolute inset-0 cursor-not-allowed flex justify-center items-center" />
      )}
    </div>
  );
};

export default TaskForm;
