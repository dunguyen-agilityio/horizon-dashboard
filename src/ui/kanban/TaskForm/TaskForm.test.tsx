import { act, fireEvent, render, waitFor } from '@testing-library/react';
import TaskForm from '.';
import { STATUS } from '@/models/Task';
import { TASK_MESSAGE } from '@/constants/messages';

const mockChangedTitle = 'Implement Task Board Component';

describe('Task Form tests', () => {
  let mockOnClose: jest.Mock;
  let mockOnSubmit: jest.Mock;

  beforeEach(() => {
    mockOnClose = jest.fn();
    mockOnSubmit = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const setup = () =>
    render(
      <TaskForm
        onClose={mockOnClose}
        onSubmit={mockOnSubmit}
        status={STATUS.BACKLOG}
      />,
    );

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should show error message when click Create button and title invalid', async () => {
    const { getByTestId, getByText, queryByText } = setup();

    act(() =>
      fireEvent.change(getByTestId('title-field'), { target: { value: 'Ne' } }),
    );

    act(() => fireEvent.click(getByText('Create task')));

    await waitFor(() => {
      expect(queryByText(TASK_MESSAGE.MIN_LENGTH)).toBeTruthy();
      expect(mockOnSubmit).not.toHaveBeenCalled();
    });
  });

  it('addNewTask will called when click Create button', async () => {
    const { getByTestId, getByText } = setup();

    act(() =>
      fireEvent.change(getByTestId('title-field'), {
        target: { value: mockChangedTitle },
      }),
    );

    act(() => fireEvent.click(getByText('Create task')));

    await waitFor(() =>
      expect(mockOnSubmit).toHaveBeenNthCalledWith(1, {
        title: mockChangedTitle,
        status: STATUS.BACKLOG,
      }),
    );
  });
});
