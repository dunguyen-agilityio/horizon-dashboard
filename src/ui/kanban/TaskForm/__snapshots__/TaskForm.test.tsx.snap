// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Form tests Should match snapshot 1`] = `
<div>
  <div
    class="sticky top-0 z-10 pt-[18px] pl-[23px] pb-[23px] pr-[21px] rounded-[15px] bg-white shadow-light-card dark:shadow-none dark:bg-indigo-light max-w-[466px]"
  >
    <form
      class="flex flex-col gap-4"
      data-testid="add-new-task-form"
    >
      <div
        class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
        data-filled="true"
        data-filled-within="true"
        data-has-elements="true"
        data-has-label="true"
        data-slot="base"
      >
        <div
          class="h-full flex flex-col"
          data-slot="main-wrapper"
        >
          <div
            class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 bg-default-100 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 h-10 min-h-10 rounded-medium transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background"
            data-slot="input-wrapper"
            style="cursor: text;"
          >
            <label
              class="absolute pointer-events-none origin-top-left rtl:origin-top-right subpixel-antialiased block text-foreground-500 will-change-auto !duration-200 !ease-out motion-reduce:transition-none transition-[transform,color,left,opacity] group-data-[filled-within=true]:text-foreground group-data-[filled-within=true]:pointer-events-auto pb-0 z-20 top-1/2 -translate-y-1/2 group-data-[filled-within=true]:left-0 left-3 rtl:left-auto rtl:right-3 text-small group-data-[filled-within=true]:-translate-y-[calc(100%_+_theme(fontSize.small)/2_+_20px)] pe-2 max-w-full text-ellipsis overflow-hidden"
              data-slot="label"
              for="react-aria-:r0:"
              id="react-aria-:r1:"
            >
              Title
            </label>
            <div
              class="inline-flex w-full items-center h-full box-border"
              data-slot="inner-wrapper"
            >
              <input
                aria-label="Title"
                aria-labelledby="react-aria-:r0: react-aria-:r1:"
                class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground"
                data-slot="input"
                data-testid="title-field"
                id="react-aria-:r0:"
                name="title"
                placeholder="Enter title"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex justify-between"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium pointer-events-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none border-default data-[hover=true]:opacity-hover text-black dark:text-white opacity-50 cursor-not-allowed bg-transparent"
          data-disabled="true"
          disabled=""
          type="submit"
        >
          Create task
        </button>
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 border-medium px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none border-default text-default-foreground data-[hover=true]:opacity-hover bg-transparent"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</div>
`;
